### ISSUE TITLE
debtap -u fails with 404 error

### ISSUE THREAD

**Issue by garyvdm** [helixarch/debtap#5]
When I try run `debtap -u` I get a 404 not found error: 

```
garyvdm@garyvdm ~ % sudo debtap -u                                                            :(
==> Synchronizing pkgfile database...
:: Updating 5 repos...
  infinality-bundle is up to date
  multilib is up to date
  core is up to date
  extra is up to date
  community is up to date
==> Synchronizing debtap database...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29.7M  100 29.7M    0     0   498k      0  0:01:01  0:01:01 --:--:--  399k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (22) The requested URL returned error: 404 Not Found

gzip: /var/cache/debtap/ubuntu-packages-files.gz: unexpected end of file
Synchronization failed. Exiting...
```

Using version 2.1.5-1

**Comment by helixarch** [comment]
I'm currently on vacations and have no access to a Linux PC. Can you please give me the output of this:

curl -s http://packages.ubuntu.com | grep option | gawk -F '=' '{print $2}' | gawk '{print $1}' | grep -v option | tac | sed -n 2'{p;q;}' | sed s'/\("\|-updated\|-backports\)//g'

**Comment by helixarch** [comment]
Ok, I asked a friend to give me the output of this comnand. The result is "wily" (correct for this period). Then he run the curl to download the file from the url that gives you error 404. Again, there was no error; it run perfectly. A possible explanation is that their server was down at the moment you run "debtap -u" and you should rerun the command.

**Comment by helixarch** [comment]
No response. I'm closing this as solved...

**Comment by xayd01** [comment]
Not so fast ;).

Same issue here.  I can also reproduce "willy" from the curl above.

[root@stratux debtap]# uname -a
Linux stratux 3.4.29+ #1 PREEMPT Sat Aug 15 20:51:17 CDT 2015 armv7l GNU/Linux
[root@stratux debtap]# ./debtap --update
==> Synchronizing pkgfile database...
:: Updating 5 repos...
  core is up to date
  extra is up to date
  alarm is up to date
  aur is up to date
  community is up to date
==> Synchronizing debtap database...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (22) The requested URL returned error: 404 Not Found

gzip: /var/cache/debtap/debian-packages-files.gz: unexpected end of file
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (22) The requested URL returned error: 404 Not Found

gzip: /var/cache/debtap/ubuntu-packages-files.gz: unexpected end of file
Synchronization failed. Exiting...

**Comment by talesam** [comment]
[neit Itaú]# debtap warsaw_setup_64.deb 
Error: You must run at least once "debtap -u" with root privileges (preferably recently), before running this script

**Comment by helixarch** [comment]
Make a commit for this:

https://github.com/helixarch/debtaparm

Remember this is ONLY for ARM architecture, so keep i386/i686/x86_86 out of this. After this, I will make a release.

**Comment by cecton** [comment]
okay thanks. Maybe you should specify the architectures manually in the PKBUILD then?

**Comment by helixarch** [comment]
I'm not using a machine

**Comment by helixarch** [comment]
I'm not using any machine with ARM architecture, if you want to create a PKGBUILD for debtaparm and upload it somewhere, feel free to do so.

**Comment by cecton** [comment]
I mean here: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=debtap

It say architecture=any but it should be i686 and x86_64.

**Comment by helixarch** [comment]
No, it's correct because it's a bash script and works on any platform (i686 and x86_64 for the original debtap), it doesn't need to be compiled for a specific platform.

**Comment by cecton** [comment]
Then maybe we can prevent the mistake from happening:

```diff
                     _arch=i386
                 elif [[ "$(uname -m)" == "x86_64" ]]; then
                     _arch=amd64
+                else
+                    echo -n "Unknown architecture: "; uname -m
+                    exit 1
                 fi
```

**Comment by helixarch** [comment]
The lines about i386/i686/amd64/x86_64 are not needed in debtaparm

**Comment by cecton** [comment]
Sorry, it slipped my mind. I was sorting out my old repositories, deleting the useless ones and I found the PR I made you a long time ago. I was recalling why and what and then I realize...

btw... there is another PR for you with someone who want the support for arm and it's ignored. Multiple PR of people confused by your program and I just don't get why you don't understand that this is confusing.

People are wondering why they get 500 or 404 errors despite they just installed a simple packet correctly and expect it to work. The reason is simple: it's not made for arm. Then you have 2 choices:

1. You make it crystal clear to people: the PKGBUILD restricts it and/or the program ensures on startup you are running the correct architecture.
2. You allows people who can test on ARM to make the upgrade for you. That's one of the perk of being open source after all.

But not doing anything is just attracting more people confused by a late error in your program that doesn't reveal the real issue. Imagine for a non-programmer what it looks like.

What I don't even less understand is that you actually have made a variant for arm but not put it on AUR. The only reason I would understand is laziness because I'm so good myself at postponing stuff to later. But when I see people confused by my products, I can't help to be willing to sort out the problems.

**Comment by helixarch** [comment]
I have made clear multiple times in both Arch Linux forums and AUR that this wasn't meant to used with ARM devices. I really want to make it work, despite the fact that I have no device to test it on, but whenever I contacted people to do some tests for me, they simply ignored me. I can even  forward to you my e-mails that show this very clearly. To make it work for ARM devices more is needed than just changing a few urls and so far I haven't got any help.

**Comment by cecton** [comment]
Ok thanks I understand that. But then why not restricting?

**Comment by helixarch** [comment]
Because my intention is not to restrict it, but work around the issue with people who are willing to do some tests for me

**Comment by cecton** [comment]
Hello,

Sorry to bother, I would like to add to things:

1. According to you:

> No, it's correct because it's a bash script and works on any platform (i686 and x86_64 for the original debtap), it doesn't need to be compiled for a specific platform.

But according to https://wiki.archlinux.org/index.php/PKGBUILD#arch
> If instead a package can be compiled for any architecture, but is architecture-specific once compiled, specify all architectures officially supported by Arch, i.e. arch=('i686' 'x86_64'). 

2. This PR looks good to me: #16 

Because (not because it's mine): a. it fails cleanly for unsupported architectures; b. it handles the download URL differently for each architecture; c. this is the official repository of Ubuntu for the arm, ppc and amd64 architectures (and some other stuff).

**Comment by helixarch** [comment]
I'll take a look it in about one month, I'm currently finishing my other project "subedit" which is what kept me mostly away last year

**Comment by cecton** [comment]
Nice project. I did a srt editor (very simple) in Perl back in the days. Now I'm wondering, why this devotion for bash? You seem very talented in bash but it's so much easier and faster to get result in Python for instance.

**Comment by helixarch** [comment]
I don't know... I just love bash

**Comment by helixarch** [comment]
As promised I am back after a month to fix this issue once and forever for the -soon to be released- new version of debtap. Any ideas?

**Comment by Foryah** [comment]
Is there a fix to this ? I'm still having this problem... Here's my output : 
```
 ~ s debtap --update                                                                         
[sudo] password for david: 
==> Synchronizing pkgfile database...
:: Updating 4 repos...
  multilib is up to date
  core is up to date
  community is up to date
  extra is up to date
==> Synchronizing debtap database...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 32.9M  100 32.9M    0     0  1692k      0  0:00:19  0:00:19 --:--:-- 5193k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:05 --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found

gzip: /var/cache/debtap/ubuntu-packages-files.gz: unexpected end of file
Synchronization failed. Exiting...
 ~ uname -m
x86_64
 ~ curl -s http://packages.ubuntu.com | grep option | gawk -F '=' '{print $2}' | gawk '{print $1}' | grep -v option | tac | sed -n 3'{p;q;}' | sed s'/\("\|-updated\|-backports\)//g'
 ~ # No output
```

**Comment by helixarch** [comment]
A fix for this has been merged today, hopefully soon I will update the AUR package of debtap

**Comment by Kalman98** [comment]
I had this error as well, I can confirm that the latest code from GitHub works for me. Thanks! 😄

**Comment by Foryah** [comment]
Same here. The latest code from GitHub work. Did not try the AUR package yet...

**Comment by helixarch** [comment]
There is some work left for next release, it brings new features

**Comment by wrathagom** [comment]
Just as a further update, cloning from github worked, but AUR still gives the above 404.

**Comment by audiomuze** [comment]
Likewise, puled from github a few minutes ago and still get a 404.

sudo debtap -u
[sudo] password for x: 
==> Synchronizing pkgfile database...
:: Updating 4 repos...
  extra is up to date
  community is up to date
  core is up to date
  multilib is up to date
==> Synchronizing debtap database...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 33.0M  100 33.0M    0     0  5166k      0  0:00:06  0:00:06 --:--:-- 5330k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found

gzip: /var/cache/debtap/ubuntu-packages-files.gz: unexpected end of file
Synchronization failed. Exiting...

**Comment by helixarch** [comment]
Use the latest patched version here until I have 3.0 ready

**Comment by audiomuze** [comment]
^^^ forgive a probably stupid question, but how does one do this?  I cloned from https://github.com/helixarch/debtap.git and get the 404?  Where/ how does one pull the patched version?

**Comment by Foryah** [comment]
I think what @helixarch  meant was to use the latest github version instead of the AUR version. For me the 404 is gone while using the latest github version... 

So, for you, this fails ? It doesn't print `zesty` ?
`curl -s https://packages.ubuntu.com | grep option | gawk -F '=' '{print $2}' | gawk '{print $1}' | grep -v option | tac | sed -n 3'{p;q;}' | sed s'/\("\|-updated\|-backports\)//g'`

**Comment by helixarch** [comment]
Fixed in 3.0. Closing...
